### PR TITLE
Mixscape reproducibility

### LINKED ISSUE #671: Reproducibility of mixscape
### Report

Hello, 

When running the lda visualisation plot as part of the mixscape workflow, I am not getting consistent results between runs, the plot looks slightly different (although very similar).
Is a random seed used at any point, and is there a way to set this so I can get reproducible results?

Code used:

ms = pt.tl.Mixscape()
ms.perturbation_signature(mdata["rna"], pert_key="perturbation", control="NT")

adata_pert = mdata["rna"].copy()
adata_pert.X = adata_pert.layers["X_pert"]

sc.pp.pca(adata_pert)
sc.pp.neighbors(adata_pert, metric="cosine")
sc.tl.umap(adata_pert)

ms.mixscape(adata=mdata["rna"], control="NT", labels="guide_allocation", layer="X_pert",
            logfc_threshold = 0.15, iter_num=100, min_de_genes=5)
ms.lda(adata=mdata["rna"], control="NT", labels="guide_allocation", layer="X_pert")
ms.plot_lda(adata=mdata["rna"], control="NT")

### Version information

-----
adjustText          1.2.0
anndata             0.10.8
matplotlib          3.9.2
mudata              0.3.1
muon                0.1.6
numpy               1.26.4
pandas              2.2.2
pertpy              0.9.4
scanpy              1.10.3
session_info        1.0.0
-----
PIL                 10.4.0
absl                NA
appnope             0.1.4
asttokens           NA
attr                24.2.0
blitzgsea           NA
certifi             2024.08.30
charset_normalizer  3.3.2
chex                0.1.86
comm                0.2.2
contextlib2         NA
cycler              0.12.1
cython_runtime      NA
dateutil            2.9.0
debugpy             1.8.5
decorator           5.1.1
decoupler           1.8.0
docrep              0.3.2
equinox             0.11.7
etils               1.9.4
executing           2.1.0
filelock            3.16.1
flax                0.9.0
fsspec              2024.9.0
h5py                3.11.0
idna                3.10
igraph              0.11.6
ipykernel           6.29.5
jax                 0.4.33
jaxlib              0.4.33
jaxopt              NA
jaxtyping           0.2.34
jedi                0.19.1
joblib              1.4.2
kiwisolver          1.4.7
lamin_utils         0.13.4
legacy_api_wrap     NA
leidenalg           0.10.2
lightning           2.4.0
lightning_fabric    2.4.0
lightning_utilities 0.11.7
lineax              0.0.5
llvmlite            0.43.0
matplotlib_inline   0.1.7
ml_collections      NA
ml_dtypes           0.5.0
mpl_toolkits        NA
mpmath              1.3.0
msgpack             1.1.0
multipledispatch    0.6.0
natsort             8.4.0
numba               0.60.0
numpyro             0.15.3
opt_einsum          v3.3.0
optax               0.2.3
ott                 0.4.8
packaging           24.1
parso               0.8.4
patsy               0.5.6
pickleshare         0.7.5
platformdirs        4.3.6
ply                 3.11
prompt_toolkit      3.0.47
psutil              6.0.0
pubchempy           1.0.4
pure_eval           0.2.3
pyarrow             17.0.0
pydev_ipython       NA
pydevconsole        NA
pydevd              2.9.5
pydevd_file_utils   NA
pydevd_plugins      NA
pydevd_tracing      NA
pygments            2.18.0
pynndescent         0.5.13
pyomo               6.8.0
pyparsing           3.1.4
pyro                1.9.1
pytorch_lightning   2.4.0
pytz                2024.2
requests            2.32.3
rich                NA
scipy               1.14.1
scvi                1.1.6.post2
seaborn             0.13.2
six                 1.16.0
sklearn             1.5.2
skmisc              0.5.1
sparsecca           0.3.1
stack_data          0.6.2
statsmodels         0.14.3
sympy               1.13.3
texttable           1.7.0
threadpoolctl       3.5.0
toolz               0.12.1
torch               2.4.1
torchgen            NA
torchmetrics        1.4.2
tornado             6.4.1
tqdm                4.66.5
traitlets           5.14.3
typeguard           NA
typing_extensions   NA
umap                0.5.6
urllib3             2.2.3
wcwidth             0.2.13
yaml                6.0.2
zmq                 26.2.0
-----
IPython             8.27.0
jupyter_client      8.6.3
jupyter_core        5.7.2
-----
Python 3.12.6 | packaged by conda-forge | (main, Sep 11 2024, 04:55:15) [Clang 17.0.6 ]
macOS-13.0.1-arm64-arm-64bit
-----

### PR DESCRIPTION
**Pull request by Lilly-May**
**PR Checklist**

-   [X] Referenced issue is linked (closes #671)
-   [X] If you've fixed a bug or added code that should be tested, add tests!
-   [X] Documentation in `docs` is updated

**Description of changes**

- As described in #671, running Mixscape multiple times can produce varying results due to the non-deterministic `GaussianMixture` model. I added a seed parameter to `pt.tl.Mixscape.mixscape`.
- In the original Mixscape implementation, there is a parameter called `n_dims`, which specifies the number of dimensions from the chosen representation (usually PCA) to be used for determining the nearest neighbors. I have added this parameter to our implementation of `pt.tl.Mixscape.perturbation_signature` and set its default value to 15, which is also the default used in the original implementation.
- I added a test for `pt.tl.Mixscape.perturbation_signature` that not only verifies that a layer has been added to the AnnData object but also checks the correctness of the computed scores
- **_Input required_**: The perturbation signature was previously calculated by subtracting the average expression of the N nearest control (NT) cells from the observed expression. But, in the original implementation, this subtraction is performed in the reverse order (see [here](https://github.com/satijalab/seurat/blob/1549dcb3075eaeac01c925c4b4bb73c73450fc50/R/mixscape.R#L1280)). This only affects the sign of the result, but aligning our calculation with the original implementation would make it easier to compare scores. I suggest adopting our calculation to the original approach, but happy to discuss this further.

### CODE REVIEW

**Comment by Zethson** [comment]
The test fixes are unrelated and I'm hoping to fix them next week. Just ping me if you want a review, please.

**Comment by Lilly-May** [comment]
Originally, I wanted to add more to this PR, but I think it’s easier to create a separate one. Feel free to review, if you have time @Zethson!

File: pertpy/tools/_mixscape.py [hunk 1]
```diff
@@ -162,11 +166,13 @@ def mixscape(
...
         pval_cutoff: float | None = 5e-2,
         perturbation_type: str | None = "KO",
+            random_state: int | None = 0,
         copy: bool | None = False,
     ):
         """Identify perturbed and non-perturbed gRNA expressing cells that accounts for multiple treatments/conditions/chemical perturbations.
 
-        The implementation resembles https://satijalab.org/seurat/reference/runmixscape
+        The implementation resembles https://satijalab.org/seurat/reference/runmixscape. Note that in the original implementation, the
+        perturbation signature is calculated on unscaled data by default and we therefore recommend to do the same.
```
Comment (Zethson) [inline pertpy/tools/_mixscape.py:175]: Should we test somewhere whether the data is unscaled and print a warning if it is? We can at least test whether the input is count or normalized data. There's code somewhere in pertpy for this I think. If not, please ping me.

**Review by Zethson — APPROVED** [review]
Great, thank you!

1. > Input required: The perturbation signature was previously calculated by subtracting the average expression of the N nearest control (NT) cells from the observed expression. But, in the original implementation, this subtraction is performed in the reverse order (see [here](https://github.com/satijalab/seurat/blob/1549dcb3075eaeac01c925c4b4bb73c73450fc50/R/mixscape.R#L1280)). This only affects the sign of the result, but aligning our calculation with the original implementation would make it easier to compare scores. I suggest adopting our calculation to the original approach, but happy to discuss this further. 

Yes, let's adopt the way they did it. Whenever we can align with them, let's do it.

2. I had to merge `main` into your branch to get the CI to run and hopefully pass. You may now also have the code and issue of https://github.com/scverse/pertpy/issues/694. I'll have a look at this ASAP but your results may now be different.
